src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalMultiphasePoromechanics_impl.hpp

```diff
@@ -0,0 +1,666 @@
+
+  real64 const maxForce = Base::complete( k, stack );
+
+  localIndex const numSupportPoints =
+    m_finiteElementSpace.template numSupportPoints< FE_TYPE >( stack.feStack );
+  integer numDisplacementDofs = numSupportPoints * numDofPerTestSupportPoint;
+
+  // Apply equation/variable change transformation(s)
+  real64 work[maxNumComponents + 1]{};
+  shiftRowsAheadByOneAndReplaceFirstRowWithColumnSum( m_numComponents, 1, stack.dLocalResidualMass_dTemperature, work );
+
+  // Step 1: assemble the derivatives of linear momentum balance wrt temperature into the global matrix
+
+  for( int localNode = 0; localNode < numSupportPoints; ++localNode )
+  {
+    for( integer dim = 0; dim < numDofPerTestSupportPoint; ++dim )
+    {
+      localIndex const dof = LvArray::integerConversion< localIndex >( stack.localRowDofIndex[numDofPerTestSupportPoint*localNode + dim] - m_dofRankOffset );
+      if( dof < 0 || dof >= m_matrix.numRows() )
+      {
+        continue;
+      }
+
+      m_matrix.template addToRowBinarySearchUnsorted< parallelDeviceAtomic >( dof,
+                                                                              &stack.localTemperatureDofIndex,
+                                                                              stack.dLocalResidualMomentum_dTemperature[numDofPerTestSupportPoint * localNode + dim],
+                                                                              1 );
+    }
+  }
+
+  localIndex const massDof = LvArray::integerConversion< localIndex >( stack.localPressureDofIndex - m_dofRankOffset );
+  if( 0 <= massDof && massDof < m_matrix.numRows() )
+  {
+
+    // Step 2: assemble the derivatives of mass balance residual wrt temperature into the global matrix
+
+    for( localIndex i = 0; i < m_numComponents; ++i )
+    {
+      m_matrix.template addToRow< serialAtomic >( massDof + i,
+                                                  &stack.localTemperatureDofIndex,
+                                                  stack.dLocalResidualMass_dTemperature[i],
+                                                  1 );
+    }
+
+    // Step 3: assemble the derivatives of pore-volume constraint residual wrt temperature into the global matrix
+
+    m_matrix.template addToRow< serialAtomic >( massDof + m_numComponents,
+                                                &stack.localTemperatureDofIndex,
+                                                stack.dLocalResidualPoreVolConstraint_dTemperature[0],
+                                                1 );
+  }
+
+  // Step 4: assemble the energy balance and its derivatives into the global matrix
+
+  localIndex const energyDof = LvArray::integerConversion< localIndex >( stack.localTemperatureDofIndex - m_dofRankOffset );
+
+  if( 0 <= energyDof && energyDof < m_matrix.numRows() )
+  {
+    m_matrix.template addToRowBinarySearchUnsorted< serialAtomic >( energyDof,
+                                                                    stack.localRowDofIndex,
```

Comment:
the stack.localRowDofIndex should be renamed as stack.localDisplacementDofIndex to make it consistents with the others for pressure, temperature and fluid components